src/core/qgsvectorlayerfeatureiterator.cpp

```diff
@@ -140,14 +140,35 @@ QgsVectorLayerFeatureIterator::QgsVectorLayerFeatureIterator( QgsVectorLayerFeat
     mRequest.setFilterRect( mFilterRect );
   }
 
+  // check whether the order by clause(s) can be delegated to the provider
+  mDelegatedOrderByToProvider = !mSource->mHasEditBuffer;
+  if ( !mRequest.orderBy().isEmpty() )
+  {
+    QSet<int> attributeIndexes;
+    const auto usedAttributeIndices = mRequest.orderBy().usedAttributeIndices( mSource->mFields );
+    for ( int attrIndex : usedAttributeIndices )
+    {
+      if ( mSource->mFields.fieldOrigin( attrIndex ) != QgsFields::OriginProvider )
+        mDelegatedOrderByToProvider = false;
+
+      attributeIndexes << attrIndex;
+    }
+
+    if ( mRequest.flags() & QgsFeatureRequest::SubsetOfAttributes )
```

Comment:
```suggestion
    if ( mRequest.flags() & QgsFeatureRequest::SubsetOfAttributes && !mDelegatedOrderByToProvider )
```

Reply:
Hm, are you sure about that?

Reply:
I guess it makes sense since we're adding those order by fields to the mProviderRequest down the line.

Reply:
Is there a chance left that we will do the order by on this level even if we delegate to provider?